### PR TITLE
Reader: add Customer Council banner

### PR DESCRIPTION
**Pull request by holdercp**
Closes https://github.com/Automattic/dotcom-forge/issues/9343

## Proposed Changes

* Add new banner to drive subscriptions to customer council P2

### Mobile

image

### Desktop

image

## Why are these changes being made?

* To gather well-rounded feedback, we’ll show a dismissible banner in the Reader’s following feed for 24 hours to start.

## Testing Instructions

* Navigate to `/read` or `/discover`
* Click "subscribe" on banner
* You should navigate to (https://readercouncilgeneral.wordpress.com/)
* You should be subscribed to the P2
* Navigate back to a Reader feed and the banner should not appear
* Repeat on mobile

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16898994

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @holdercp for including a screenshot in the description! This is really helpful for our translators.

File: client/reader/stream/index.jsx [hunk 1]
```diff
@@ -665,6 +667,9 @@ class ReaderStream extends Component {
 				body = (
 					<>
 						{ streamHeader?.() }
+						<div style={ { margin: '32px 16px 0' } }>
```
Comment (holdercp) [inline client/reader/stream/index.jsx]: Creating space on mobile; didn't want to add another stylesheet since this is temporary.

File: client/reader/stream/customer-council-banner.jsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+			onClick={ subscribeToP2AndNavigate }
+			description={ translate(
+				'Join {{a}}our new blog{{/a}} to share your feedback and help us improve your reading experience.',
+				{
+					components: {
+						a: <a href={ CUSTOMER_COUNCIL_P2_URL } />,
+					},
+				}
+			) }
+		/>
```
Comment (holdercp) [inline client/reader/stream/customer-council-banner.jsx]: There are other props related to tracking we can pass to this component. I'm not sure if we want to leverage any of them: https://wpcalypso.wordpress.com/devdocs/design/banner

File: client/reader/stream/customer-council-banner.jsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+	if ( hideBanner ) {
+		return null;
+	}
+
+	const subscribeToP2AndNavigate = () => {
+		subscribe(
+			{ blog_id: CUSTOMER_COUNCIL_P2_URL, url: CUSTOMER_COUNCIL_P2_URL },
+			{
+				onSettled: () => {
```
Comment (holdercp) [inline client/reader/stream/customer-council-banner.jsx:36]: We're navigating regardless of if the subscription succeeded or not.

**Comment by holdercp** [comment]
@Automattic/loop I added the banner to the feeds in `/discover` and `/read` feeds but not on `/search`, `/likes`, etc.

**Comment by davemart-in** [comment]
Good progress @holdercp. A couple of things:

1. Can we limit the banner to just showing up on /read and /discover? I'm currently also seeing it on tag pages, P2 pages, and Automattic P2 pages. Can we remove it from those pages?
2. When I click the "Subscribe" button I am subscribed to the blog, but I'm redirected to `plans/{SITE-NAME}`, not https://readercouncilgeneral.wordpress.com/

File: client/reader/stream/customer-council-banner.jsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+	const hideBanner = ( alreadySubscribed && checkedAlreadySubscribed ) || checkingAlreadySubscribed;
+
+	if ( hideBanner ) {
+		return null;
+	}
+
+	const subscribeToP2AndNavigate = () => {
+		subscribe(
+			{ blog_id: CUSTOMER_COUNCIL_P2_URL, url: CUSTOMER_COUNCIL_P2_URL },
```
Comment (davemart-in) [inline client/reader/stream/customer-council-banner.jsx]: Should: `blog_id: CUSTOMER_COUNCIL_P2_URL` be `blog_id: CUSTOMER_COUNCIL_P2_ID`?

Comment (holdercp) [reply client/reader/stream/customer-council-banner.jsx]: Yes! Thank you!

**Review by davemart-in — APPROVED** [review]
Local and Calypso Live both work as expected now. Nice job tracking that issue down!

File: client/reader/stream/index.jsx [hunk 5]
```diff
@@ -601,6 +602,8 @@ class ReaderStream extends Component {
 		const { translate, forcePlaceholders, lastPage, streamHeader, streamKey, selectedPostKey } =
 			this.props;
 		const wideDisplay = this.props.width > WIDE_DISPLAY_CUTOFF;
+		const showCustomerCouncilBanner =
+			this.props.isDiscoverStream || this.props.streamKey === 'following';
```
Comment (Addison-Stavlo) [inline client/reader/stream/index.jsx]: Very minor nitpick.

But this logic doesn't determine if the banner actually shows up (since we have the hideBanner check in the banner), we could call it something like `maybeShow...` or `isCouncilPromoStream` or something else along those lines. That said, very minor and this will work perfectly fine the way it is.

File: client/reader/stream/customer-council-banner.jsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import { Reader, SubscriptionManager } from '@automattic/data-stores';
+import Banner from 'calypso/components/banner';
+import { navigate } from 'calypso/lib/navigate';
+
+export const CustomerCouncilBanner = ( { translate } ) => {
+	const CUSTOMER_COUNCIL_P2_URL = 'https://readercouncilgeneral.wordpress.com/';
+	const CUSTOMER_COUNCIL_P2_ID = '237686330';
```
Comment (Addison-Stavlo) [inline client/reader/stream/customer-council-banner.jsx]: Note - Being constants like this we could move them out of the component (have them live between imports and the banner component) so they can be exported as well. Although, we don't currently have any need for that and can always move these in another diff if we do end up having that need. Anyways, this is totally fine right now. 😄

File: client/reader/stream/customer-council-banner.jsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		SubscriptionManager.useSiteSubscribeMutation();
+
+	const {
+		data: p2,
+		isFetched: checkedAlreadySubscribed,
+		isFetching: checkingAlreadySubscribed,
+	} = Reader.useReadFeedSiteQuery( Number( CUSTOMER_COUNCIL_P2_ID ) );
+	const alreadySubscribed = p2?.is_following;
+
+	const hideBanner = ( alreadySubscribed && checkedAlreadySubscribed ) || checkingAlreadySubscribed;
```
Comment (Addison-Stavlo) [inline client/reader/stream/customer-council-banner.jsx]: Not a critique on your PR. But I am very interested to see how much response we get by showing this to all users at once! 

If we did have a desire to limit how many users see this at first, we could check some variation of `userId % 100 < X` where X is the percent we do or don't want to show it to, and ramp up the number as needed.  I think we had that idea initially in January before we got pulled away from this but have decided to just show this to everyone for a shorter amount of time instead?

Long story short, I'm just interested to see the response from this! 😅

**Review by Addison-Stavlo — APPROVED** [review]
This looks good to me, thanks @holdercp !

A couple notes, but nothing blocking from my end. 🚀

File: client/reader/stream/customer-council-banner.jsx (hunk 6 above)
Comment (holdercp) [reply client/reader/stream/customer-council-banner.jsx]: Duh! 🤦‍♂️. I originally had this component inlined in the ReaderStream component file while iterating and defined these in the component to keep things tidy but I overlooked moving them out when I move this to its own file. Moving them will also keep them from being redefined each render too, so thanks for the callout!

File: client/reader/stream/index.jsx (hunk 5 above)
Comment (holdercp) [reply client/reader/stream/index.jsx]: Yeah, great callout. I like `isCouncilPromoStream`.

File: client/reader/stream/customer-council-banner.jsx (hunk 7 above)
Comment (holdercp) [reply client/reader/stream/customer-council-banner.jsx]: Do we have a mechanism/infra in place to do slow rollouts like you're describing?

Comment (Addison-Stavlo) [reply client/reader/stream/customer-council-banner.jsx]: Not off the top of my head. We would probably do a check using something like `userId % 100 < X`, and just incrementally push PRs to update the value of X to include more users. If we really wanted we could set that number to be based on date, fetched from an endpoint, or there are probably some other options as well in that kind of case.

I think I was only involved in 1 or 2 "slow rollouts" like this and we just updated the value in later PRs when we wanted to increase the number.

**Comment by davemart-in** [comment]
@holdercp I retested this and it works as expected. It's hidden when logged in as an Automattician and it shows when I'm not. Ship it!

**Comment by holdercp** [comment]
Waiting for translation. I'll ship once it gets to ~70% complete.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16898994).
